WIP: Add quirk for Sonoff TRVZB

jayme-github opened the pull request:
This only adds child lock and window open attributes from the Sonoff manufacturer specific cluster without any further integration.

I've zero experience in Zigbee/ZHA/quirk space and was just playing around a bit. Maybe this can be a starting point but I feel like there are multiple things I don't yet understand how they're supposed to be implemented/quirked. So help/comments/discussions appreciated!

/cc @TheJulianJES  as of https://github.com/home-assistant/core/pull/105453
There is an open issue regarding this device at https://github.com/zigpy/zha-device-handlers/issues/2772

## Proposed change
<!--
  Explain your proposed change below.
-->


## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

jayme-github committed:
Add quirk for Sonoff TRVZB

This only adds child lock and window open attributes from the Sonoff
manufacturer specific cluster without any further integration.

TheJulianJES commented on zhaquirks/sonoff/trvzb.py:
```diff
@@ -0,0 +1,116 @@
+            id=SONOFF_ATTR_WINDOW_OPEN,
+            type=t.Bool,
+            mandatory=True,
+            is_manufacturer_specific=False,
+        )
+
+    @property
+    def _is_manuf_specific(self) -> bool:
+        """Pretend this is not manufacturer specific as the device will always
+        return Status.UNSUPPORTED_ATTRIBUTE if the manufacturer id is set
+        """
+        return False
```
So both changing `is_manufacturer_specific` to `False` for the attribute and overriding `_is_manuf_specific` (which probably shouldn't be done) still causes zigpy/ZHA to send the manufacturer attribute code thus causing the Sonoff to return "unsupported attribute"?

jayme-github replied:
Yes, exactly. Reading he child lock via HA's "Manage Zigbee decive" dialog does still set `is_manufacturer_specific=True` as well as `manufacturer=4742`:

```
2023-12-13 09:31:13.840 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=True, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True), manufacturer=4742, tsn=15, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>)
2023-12-13 09:31:13.840 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request: Read_Attributes(attribute_ids=[0])
2023-12-13 09:31:13.841 DEBUG (MainThread) [zigpy.device] [0xd636] Extending timeout for 0x0f request
2023-12-13 09:31:15.550 DEBUG (MainThread) [zigpy.application] Received a packet: ZigbeePacket(timestamp=datetime.datetime(2023, 12, 13, 8, 31, 15, 550115, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0xD636), src_ep=1, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), dst_ep=1, source_route=None, extended_timeout=False, tsn=0, profile_id=260, cluster_id=64529, data=Serialized[b'\x1c\x86\x12\x0f\x01\x00\x00\x86'], tx_options=<TransmitOptions.NONE: 0>, radius=29, non_member_radius=0, lqi=196, rssi=None)
2023-12-13 09:31:15.551 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Received ZCL frame: b'\x1c\x86\x12\x0f\x01\x00\x00\x86'
2023-12-13 09:31:15.551 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=True, direction=<Direction.Client_to_Server: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True), manufacturer=4742, tsn=15, command_id=1, *direction=<Direction.Client_to_Server: 1>)
2023-12-13 09:31:15.552 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame: SonoffManufCluster:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0000, status=<Status.UNSUPPORTED_ATTRIBUTE: 134>)])
2023-12-13 09:31:15.553 DEBUG (MainThread) [homeassistant.components.zha.websocket_api] Read attribute for: cluster_id: [64529] cluster_type: [in] endpoint_id: [1] attribute: [0] manufacturer: [4742] response: [None] failure: [{0: <Status.UNSUPPORTED_ATTRIBUTE: 134>}],
```


vs:

```
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=False, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True), tsn=16, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>)
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request: Read_Attributes(attribute_ids=[0])
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.device] [0xd636] Extending timeout for 0x10 request
2023-12-13 09:34:32.267 DEBUG (MainThread) [zigpy.application] Received a packet: ZigbeePacket(timestamp=datetime.datetime(2023, 12, 13, 8, 34, 32, 267321, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0xD636), src_ep=1, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), dst_ep=1, source_route=None, extended_timeout=False, tsn=0, profile_id=260, cluster_id=64529, data=Serialized[b'\x18\x10\x01\x00\x00\x00\x10\x00'], tx_options=<TransmitOptions.NONE: 0>, radius=29, non_member_radius=0, lqi=196, rssi=None)
2023-12-13 09:34:32.267 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Received ZCL frame: b'\x18\x10\x01\x00\x00\x00\x10\x00'
2023-12-13 09:34:32.268 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=0, direction=<Direction.Client_to_Server: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True), tsn=16, command_id=1, *direction=<Direction.Client_to_Server: 1>)
2023-12-13 09:34:32.268 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame: SonoffManufCluster:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0000, status=<Status.SUCCESS: 0>, value=TypeValue(type=Bool, value=<Bool.false: 0>))])
2023-12-13 09:34:32.269 DEBUG (MainThread) [homeassistant.components.zha.websocket_api] Read attribute for: cluster_id: [64529] cluster_type: [in] endpoint_id: [1] attribute: [0] manufacturer: [None] response: [Bool.false] failure: [{}],
```

jayme-github replied:
> overriding `_is_manuf_specific` (which probably shouldn't be done) ...

AIUI `_has_manuf_arttr` will always return true as long as `_is_manuf_specific` is true (without checking the actual attributes). That's why I've tried it that way.

KaiDroste commented:
Hi,
I have tried your quirk. For me it is not working. With your Quirk I am not able to change values for the TVRZB. Your values were also not displayed. What makes me wonder is that for Clusters for Childmode and windowmode are not available in the cluster browser. 
Since this is the first time, I am experimenting with clusters it could be the case that i have made something wrong.

KaiDroste commented:
Hi,
I have tryed your quirk. For me it is not working. With your Quirk I am not able to change values for the TVRZB. Your values were also not displayed. What makes me wonder is that for Clusters for Childmode and windowmode are not available in the cluster browser. 


Since this is the first time, I am experimenting with ZHA quirks it could be the case that i have made something wrong. 

this is the .json file from my TRVZB: [zha-704d08d7027909b328ce78dead5adde6-SONOFF TRVZB-6826b3c1b9b1842e4077cf02615ed158.json.txt](https://github.com/zigpy/zha-device-handlers/files/13729950/zha-704d08d7027909b328ce78dead5adde6-SONOFF.TRVZB-6826b3c1b9b1842e4077cf02615ed158.json.txt)

jayme-github committed:
Add child_lock, open_window, frost_protection_temperature

This adds the actual LocalDataCluster's required to manage these
attributes from within ZHA.

jayme-github commented:
> Hi, I have tryed your quirk. For me it is not working. With your Quirk I am not able to change values for the TVRZB. Your values were also not displayed. What makes me wonder is that for Clusters for Childmode and windowmode are not available in the cluster browser.

There where no additional clusters defined in the initial commit and yes, reading/writing those attributes via the Manufacturer specific cluster is not currently possible (that's what https://github.com/home-assistant/core/pull/105453 is about).

jayme-github commented:
@TheJulianJES I'd love to get some feedback on this as I still don't feel very confident in the code base. Happy to add tests if this goes into the right direction.

The `local_temperature_calibration` will come with https://github.com/home-assistant/core/pull/105765

TheJulianJES commented:
Still on my list to look at this, but also check out: https://github.com/zigpy/zha-device-handlers/pull/2916
It seems to be the same issue you're facing.

The approach also isn't ideal IMO, but might be fine for now.

keberl commented:
> @TheJulianJES I'd love to get some feedback on this as I still don't feel very confident in the code base. Happy to add tests if this goes into the right direction.
> 
> The `local_temperature_calibration` will come with [home-assistant/core#105765](https://github.com/home-assistant/core/pull/105765)

I really appreciate this feature, but calibration values from -2.5 to +2.5 are too small.

The Sonoff TRVZB supports values from -12.8 to +12,7 °C. 
(-128 to 127 are valid values for the zigbee-attribute 'local_temperature_calibration')

Unfortunately i have no experience with quirks for zha, but is it possible to extend the calibration range?

TheNoim commented:
> > @TheJulianJES I'd love to get some feedback on this as I still don't feel very confident in the code base. Happy to add tests if this goes into the right direction.
> > The `local_temperature_calibration` will come with [home-assistant/core#105765](https://github.com/home-assistant/core/pull/105765)
> 
> I really appreciate this feature, but calibration values from -2.5 to +2.5 are too small.
> 
> The Sonoff TRVZB supports values from -12.8 to +12,7 °C. (-128 to 127 are valid values for the zigbee-attribute 'local_temperature_calibration')
> 
> Unfortunately i have no experience with quirks for zha, but is it possible to extend the calibration range?

Yes, this is waaaay to small. For example:

![image](https://github.com/zigpy/zha-device-handlers/assets/5133056/57b8cc4d-b27c-4192-86a6-8751be129acf)

But the actual temperature is:

![image](https://github.com/zigpy/zha-device-handlers/assets/5133056/7f6e04b3-b913-4cd8-b26f-25cc8dc99089)

This would require a value of -4.8

SuperSandro2000 commented:
Are there any plans to finish this?

jayme-github commented:
> Are there any plans to finish this?

I'm using the quirk as is since the initial MR. I can give it another look to see if removing the `_is_manuf_specific` override works with the changes to HA in a couple of days. Maybe it gets accepted then.

fgsch commented:
I'm also interested in seeing some of the functionality in this PR merged. Anything I can do to help?

fgsch commented:
In the meantime, I've opened https://github.com/zigpy/zha-device-handlers/pull/3358 as an alternative to this.

TheJulianJES commented:
We've merged this v2 quirk PR now: https://github.com/zigpy/zha-device-handlers/pull/3358
@jayme-github Still, thanks for your PR and sorry for the long wait time.
